### PR TITLE
Ammonia aron kozak

### PR DESCRIPTION
**Pull request by kozakaron**

### CODE REVIEW

File: full_bubble_model.py [hunk 1]
```diff
@@ -47,22 +47,34 @@
...
+except Exception as _error:
     print(print(colored('Error, \'parameters.py\' not found','red')))
+    raise _error
+
+# import excitation.py as excitation:
 try:
     import excitation
     importlib.reload(excitation)
-except:
+except ImportError as _error:
```
Comment (kozakaron) [inline full_bubble_model.py:64]: Better error handling, when importing:

So far, the role of this try-except block was to import either excitation or Bubble_synamics_simulation.excitation appropriately.

From now on, error messages won't be lost (e.g. numpy is not installed, syntax error in the file)

File: full_bubble_model.py [hunk 2]
```diff
@@ -103,6 +116,84 @@ def _colorTF(boolean):
 print(f'enable heat transfer: {_colorTF(enable_heat_transfer)}\tenable evaporation: {_colorTF(enable_evaporation)}\tenable reactions: {_colorTF(enable_reactions)}\tenable dissipated energy: {_colorTF(enable_dissipated_energy)}')
 if target_specie not in par.species:
     print(colored(f'Error, target specie \'{target_specie}\' not found in parameters.py', 'red'))
+
+
+"""________________________________Control parameters________________________________"""
+
+cpar_rules = dict(
```
Comment (kozakaron) [inline full_bubble_model.py:123]: The existance, type and value of control parameters (in cpar) is now properly checked. A correct cpar is now enforced, no default values (e.g. when mu_L is missing). Proper error messages

File: full_bubble_model.py [hunk 3]
```diff
@@ -179,25 +253,6 @@ def _initial_condition(cpar, evaporation=False, extra_dims=0):
      * lowpressure_warning: if True, the pressure during the expansion is lower, than the saturated water pressure
     """
 
-    if type(cpar) == dict:
```
Comment (kozakaron) [inline full_bubble_model.py:182]: Redundant checking and default values removed

File: full_bubble_model.py [hunk 4]
```diff
@@ -577,21 +624,14 @@ def solve(cpar, t_int=np.array([0.0, 1.0]), LSODA_timeout=30.0, Radau_timeout=30
...
-    if not 'P_v' in cpar:
-        cpar.P_v = vapour_pressure(T=cpar.T_inf) # [Pa]
-    if not 'mu_L' in cpar:
-        cpar.mu_L = viscosity(T=cpar.T_inf) # [Pa]
-    if not 'rho_L' in cpar:
-        cpar.rho_L = par.rho_L # [kg/m^3]
-    if not 'c_L' in cpar:
-        cpar.c_L = par.c_L # [m/s]
+    num_sol = None
+    if not check_cpar(cpar):
```
Comment (kozakaron) [inline full_bubble_model.py:628]: solve() can print error message and call stack when a fatal error is encountered

Python can't trace back errors in JIT-ted functions, so remove @jit when hunting for errors)

File: full_bubble_model.py [hunk 5]
```diff
@@ -941,17 +982,17 @@ def plot(cpar, t_int=np.array([0.0, 1.0]), n=5.0, base_name='', format='png', LS
     if type(cpar) == dict:
         cpar = dotdict(cpar)
 
-    num_sol, error_code, elapsed_time = solve(cpar, t_int, LSODA_timeout, Radau_timeout, extra_dims)
+    num_sol, error_code, elapsed_time = solve(cpar, t_int, LSODA_timeout, Radau_timeout, extra_dims, print_errors=True)
     data = get_data(cpar, num_sol, error_code, elapsed_time)
     
 # Print errors
     errors, success = get_errors(error_code, printit=True)
-    if not success:
```
Comment (kozakaron) [inline full_bubble_model.py:949]: from now on, plot() will draw a graph until the last succesful step if convergence failiure is encountered
error messages are printed during fatal errors

File: full_bubble_model.py (hunk 1 above)
Comment (kozakaron) [reply full_bubble_model.py:64]: same for all other imports

**Review by mrkf100 — APPROVED** [review]

**Pull request merged**